### PR TITLE
Initial unet tests

### LINKED ISSUE #97: Initial UNET++ tests

### PR DESCRIPTION
**Pull request by martynkaqhe**
## Motivation
The goal of this PR is to significantly boost the code coverage for the AI pipeline, specifically targeting `ai/unet/Model Training.py`, to help the overall repository safely exceed the 60-70% codecov requirement. 

## Key Changes
Implemented a comprehensive suite of unit tests focusing on data loaders, loss functions, transforms, and edge cases. The coverage for this critical file has now reached **81%** without overloading the CI/CD runners.

* **Data Loaders & Masks:** Tested format conversions, `normalize_merged_mask`, `_mask_to_tensor`, and morphological post-processing.
* **Edge Cases:** Added coverage for `FileNotFoundError` handling, missing masks warnings, and empty directory exceptions.
* **Losses & Schedulers:** Validated the initialization of all supported loss functions (`weighted_dice_focal`, `dice_focal`, `tversky`) and the learning rate scheduler.

## Testing Strategy: "Mini Training Epoch"
To achieve >80% coverage without burning GitHub Actions CPU minutes or using complex mocks, a `test_real_mini_training_epoch` was introduced:
* Dynamically generates a temporary dataset of three **16x16 pixel** dummy images.
* Runs a complete, authentic `train_single_model` loop for exactly **1 epoch**.
* Verifies the successful creation of the `.pth` model file.
* **Windows Compatibility Fix:** Temporarily overrides `DataLoader(num_workers=0)` within the test scope to prevent `_pickle.PicklingError` on Windows environments, ensuring the test is fully cross-platform and runnable locally by all developers.

## Coverage Impact
* `ai/unet/Model Training.py` coverage increased to **81%**.

Coderabbit's summary:

- Adds comprehensive unit test suite for `ai/unet/Model Training.py` with 16 new test methods in `TestUNetTraining` class
- Covers data loaders: `GlaucomaDatasetMergedMask` and `GlaucomaDatasetSeparateMasks` with shape and content validation
- Covers mask handling: `normalize_merged_mask`, `_mask_to_tensor`, and `morphological_postprocess` functions with various input formats
- Covers loss functions and schedulers: validates initialization of `weighted_dice_focal`, `dice_focal`, `tversky`, and learning rate scheduler
- Includes edge case testing: `FileNotFoundError` for broken masks, `ValueError` for empty directories, and missing-path branch in `build_datasets`
- Adds integration-style test (`test_real_mini_training_epoch`) that runs a single training epoch with temporary directories, verifies `.pth` checkpoint creation, and forces `DataLoader(num_workers=0)` for Windows compatibility
- Introduces shared test fixtures in `tests/conftest.py`: `dummy_arrays` (predefined NumPy arrays and torch tensors) and `dummy_dataset` (temporary directory structure with image/mask files)
- Increases code coverage for targeted file to 81% without heavy CI resource usage
- Refactors shared test setup and assertions per reviewer suggestions

[Review Change Stack](https://app.coderabbit.ai/change-stack/kequel/jaskra/pull/101?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by Agnesix — DISMISSED** [review]

File: tests/ai_unet/test_model_training.py [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+        """
+        logits = torch.tensor([[[[1.0, 1.0], [-1.0, -1.0]], 
+                                [[1.0, -1.0], [-1.0, -1.0]]]])
+        targets = torch.tensor([[[[1.0, 1.0], [0.0, 0.0]], 
+                                 [[1.0, 1.0], [0.0, 0.0]]]])
+        
+        d_score, c_score, _mean_score = model_training.dice_scores(logits, targets)
+        
+        assert abs(d_score - 1.0) < 1e-4
+        assert abs(c_score - 0.6666) < 1e-3
```
Comment (kequel) [inline tests/ai_unet/test_model_training.py]: Maybe you could use self.assert_is_ratio from BaseTest for both scores

File: tests/ai_unet/test_model_training.py [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+
+    def test_morphological_postprocess(self):
+        """
+        TEST NAME: test_morphological_postprocess
+        COMPONENT: morphological_postprocess
+        1. Create noisy dummy binary arrays for disc and cup.
+        2. Call morphological_postprocess().
+        3. Assert the output arrays have noise removed and constraints applied.
+        """
+        pred_disc = np.zeros((20, 20), dtype=np.uint8)
```
Comment (kequel) [inline tests/ai_unet/test_model_training.py]: If you could move creating any dummy arrays to conftest.py please

**Review by kequel — CHANGES_REQUESTED** [review]

**Comment by kequel** [comment]
You can also check if there is any other posibillity to move stuff to confets/base_test methods

File: tests/ai_unet/test_model_training.py (hunk 1 above)
Comment (martynkaqhe) [reply tests/ai_unet/test_model_training.py]: done

File: tests/ai_unet/test_model_training.py (hunk 2 above)
Comment (martynkaqhe) [reply tests/ai_unet/test_model_training.py]: done

**Review by kequel — APPROVED** [review]
I like it ☺️

**Review by MateuszDobry — APPROVED** [review]
Good job
